website/addons/mendeley/model.py

```diff
@@ -0,0 +1,270 @@
+import time
+
+import mendeley
+from modularodm import fields
+
+
+from website import settings
+from website.addons.base import AddonNodeSettingsBase
+from website.addons.base import AddonUserSettingsBase
+from website.citations.models import Citation
+from website.citations.models import CitationList
+from website.oauth.models import ExternalProvider
+
+from .api import APISession
+
+
+class AddonMendeleyUserSettings(AddonUserSettingsBase):
+
+    def _get_connected_accounts(self):
+        """Get user's connected Mendeley accounts"""
+        return [
+            x for x in self.owner.external_accounts if x.provider == 'mendeley'
+        ]
+
+    def to_json(self, user):
+        rv = super(AddonMendeleyUserSettings, self).to_json(user)
+        rv['accounts'] = [
+            {
+                'id': account._id,
+                'provider_id': account.provider_id,
+                'display_name': account.display_name,
+            } for account in self._get_connected_accounts()
+        ]
+        return rv
+
+
+class AddonMendeleyNodeSettings(AddonNodeSettingsBase):
+    external_account = fields.ForeignField('externalaccount',
+                                           backref='connected')
+
+    mendeley_list_id = fields.StringField()
+
+    # Keep track of all user settings that have been associated with this
+    #   instance. This is so OAuth grants can be checked, even if the grant is
+    #   not currently being used.
+    associated_user_settings = fields.AbstractForeignField(list=True)
+
+    _api = None
+
+    @property
+    def api(self):
+        """authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = Mendeley()
+            self._api.account = self.external_account
+        return self._api
+
+    def grant_oauth_access(self, user, external_account, metadata=None):
+        """Grant OAuth access, updates metadata on user settings
+
+        :param User user:
+        :param ExternalAccount external_account:
+        :param dict metadata:
+        """
+        user_settings = user.get_addon('mendeley')
+
+        # associate the user settings with this node's settings
+        if user_settings not in self.associated_user_settings:
```

Comment:
We're appending to `self.associated_user_settings` without calling `self.save()`. I'm guessing this works because something else is saving `self` indirectly, but it might be better to call it explicitly, or explain in a comment why we don't have to.